### ISSUE TITLE
Issue about the negative values in Eq. (22)

### ISSUE THREAD

**Issue by Dmsw** [xypeng9903/k-diffusion-inverse-problems#1]
How could you assure Eq. (22) resulting non-negative values?
I have used the pretraining model of guided-diffusion (https://github.com/openai/guided-diffusion.git) and found that the variance $\Sigma_\theta$ predict by neural network is always less than $\tilde\beta$ which will lead to negative values in Eq. (22).

**Comment by xypeng9903** [comment]
Yes, the prediction of posterior variance obtained using Eq. (22) sometimes (not always, do you implement it correctly?) will become negative due to the imperfect model. I address this by clipping the minimum value of the prediction by 1e-6. It is worthy to mention that Eq. (22) can be effective only for small t (for details, please refer to Section 5.1.).

**Comment by xypeng9903** [comment]
> How could you assure Eq. (22) resulting non-negative values? I have used the pretraining model of guided-diffusion (https://github.com/openai/guided-diffusion.git) and found that the variance Σθ predict by neural network is always less than β~ which will lead to negative values in Eq. (22).

**Comment by xypeng9903** [comment]
> How could you assure Eq. (22) resulting non-negative values? I have used the pretraining model of guided-diffusion (https://github.com/openai/guided-diffusion.git) and found that the variance Σθ predict by neural network is always less than β~ which will lead to negative values in Eq. (22).

Yes, the prediction of posterior variance obtained using Eq. (22) sometimes (not always, do you implement it correctly?) will become negative due to the imperfect model. I address this by clipping the minimum value of the prediction by 1e-6. It is worthy to mention that Eq. (22) can be effective only for small t (for details, please refer to Section 5.1.).

**Comment by Dmsw** [comment]
From my perspective, if we re-parameterize the network and predict the variance of the $\epsilon$ rather than directly predicting the variance of $x_0$, we can heroically derive the limitation of $\sigma_t \nabla_{x_t}\log p(y|x_t)$ when $t\rightarrow 0$. 

Specifically, we can train a neural network $\sigma_\phi$ using the following loss function:
$$L = \mathbb E_{x_0, \epsilon, t} ((\epsilon - \epsilon_\theta)^2 - \sigma_\phi (x_t, t))^2 $$
and we will have,

$$ D[x_0|x_t] = \frac{1-\bar\alpha_t}{\bar\alpha_t} D[\epsilon_\theta (x_t,t)] =  \frac{1-\bar\alpha_t}{\bar\alpha_t} \sigma_\phi ^2 (x_t, t)$$

In the case of DDPM, the incremental caused by the likelihood is

$$\Delta = \frac{1}{2}\frac{1-\alpha_t}{\sqrt{\alpha_t}}\frac{\bar\alpha_t}{1-\bar\alpha_t}\frac{\partial \hat x_{0|t}}{\partial x_t}A^T{(A\sigma_\phi^2A^T)^{-1}}(y-A\hat x_{0|t})$$

When $t\rightarrow 0$, the loss function of DM will be:

$$ L_0 = \mathbb E_{x_0, \epsilon} ||\epsilon - \epsilon_\theta(x_0, t)|_2^2|| $$

We can not predict the gaussian noise $\epsilon$ when $t\rightarrow 0$, since the gaussian noise $\epsilon$ is absolutely missing in the $x_0$ . Therefore, we theoretically have $\sigma_\phi(x_0, 0) = I$ and the network $\epsilon_\theta(x_0, 0)$ should output all zero. 

Since $$\lim_{t\rightarrow 0}\frac{1-\alpha_t}{\sqrt{\alpha_t}}\frac{\bar\alpha_t}{1-\bar\alpha_t} = 1,$$
and
$$\lim_{t\rightarrow0}\frac{\partial \hat x_{0|t}}{\partial x_t} = I$$, we have

$$\lim_{t\rightarrow 0}\Delta = \frac{\partial}{\partial x_t}||y-A\hat x_{0|t}||_2^2 = A^T(AA^T)^{-1}(Ax-y)$$

According to my experimental results, it will have a more robust and better performance in the noiseless cases. Thus, as far as I am concerned, it is incorrect to simply clip it to a small number.

**Comment by Dmsw** [comment]
> > How could you assure Eq. (22) resulting non-negative values? I have used the pretraining model of guided-diffusion (https://github.com/openai/guided-diffusion.git) and found that the variance Σθ predict by neural network is always less than β~ which will lead to negative values in Eq. (22).
> 
> Yes, the prediction of posterior variance obtained using Eq. (22) sometimes (not always, do you implement it correctly?) will become negative due to the imperfect model. I address this by clipping the minimum value of the prediction by 1e-6. It is worthy to mention that Eq. (22) can be effective only for small t (for details, please refer to Section 5.1.).

From my perspective, if we re-parameterize the network and predict the variance of the $\epsilon$ rather than directly predicting the variance of $x_0$, we can theoretically derive the limitation of $\sigma_t \nabla_{x_t}\log p(y|x_t)$ when $t\rightarrow 0$. 

Specifically, we can train a neural network $\sigma_\phi$ using the following loss function:
$$L = \mathbb E_{x_0, \epsilon, t} ((\epsilon - \epsilon_\theta)^2 - \sigma_\phi (x_t, t))^2 $$
and we will have,

$$ D[x_0|x_t] = \frac{1-\bar\alpha_t}{\bar\alpha_t} D[\epsilon_\theta (x_t,t)] =  \frac{1-\bar\alpha_t}{\bar\alpha_t} \sigma_\phi ^2 (x_t, t)$$

In the case of DDPM, the incremental caused by the likelihood is

$$\Delta =\frac{1-\alpha_t}{\sqrt{\alpha_t}}\frac{\bar\alpha_t}{1-\bar\alpha_t}\frac{\partial \hat x_{0|t}}{\partial x_t}A^T{(A\sigma_\phi^2A^T)^{-1}}(y-A\hat x_{0|t})$$

When $t\rightarrow 0$, the loss function of DM will be:

$$ L_0 = \mathbb E_{x_0, \epsilon} ||\epsilon - \epsilon_\theta(x_0, t)|_2^2|| $$

We can not predict the gaussian noise $\epsilon$ when $t\rightarrow 0$, since the gaussian noise $\epsilon$ is absolutely missing in the $x_0$ . Therefore, we theoretically have $\sigma_\phi(x_0, 0) = I$ and the network $\epsilon_\theta(x_0, 0)$ should output all zero. 

Since $$\lim_{t\rightarrow 0}\frac{1-\alpha_t}{\sqrt{\alpha_t}}\frac{\bar\alpha_t}{1-\bar\alpha_t} = 1,$$
and
$$\lim_{t\rightarrow0}\frac{\partial \hat x_{0|t}}{\partial x_t} = I$$, we have

$$\lim_{t\rightarrow 0}\Delta = \frac{\partial}{\partial x_t}||y-A\hat x_{0|t}||_2^2 = A^T(AA^T)^{-1}(Ax-y)$$

According to my experimental results, it will have a more robust and better performance in the noiseless cases. Thus, as far as I am concerned, it is incorrect to simply clip it to a small number.

Exclude this point, I think you have done quite a decent and well job. Thanks.
